### ISSUE TITLE
How to configure boss to self hosting gitlab?

### ISSUE THREAD

**Issue by No6-start** [chaoss/grimoirelab#413]
How to configure boss to self hosting gitlab? I can't understand the official tutorial. I can't successfully create my gitlab configuration file and import it

**Comment by No6-start** [comment]
And will the Chinese interface be launched？

**Comment by vchrombie** [comment]
Hi @No6-start, thanks for your interest in GrimoireLab.

> How to configure boss to self hosting gitlab? I can't understand the official tutorial. I can't successfully create my gitlab configuration file and import it

I hope you are asking about configuring a private gitlab instance for grimoirelab. You can find the whole documentation related to the configurations here, [chaoss/grimoirelab-sirmordred](https://github.com/chaoss/grimoirelab-sirmordred#contents).

I'm sharing an example configuration, please change the required fields. 

If a given GitLab repository is under more than 1 level, all the slashes `/` starting from the second level have to be replaced by `%2F`.
projects.json
```
{
    "Chaoss": {
        "gitlab:issue": [
            "https://gitlab.com/Molly/first",
            "https://gitlab.com/Molly/lab%2Fsecond"
        ],
        "gitlab:merge": [
            "https://gitlab.com/Molly/first",
            "https://gitlab.com/Molly/lab%2Fsecond"
        ]
    }
}
```

Ideally, you need to use `enterprise-url` for collecting data from private gitlab instances and also the `api-token` should be generated by a member of that project.
setup.cfg
```
[panels]
gitlab-issues = true
gitlab-merges = true

[gitlab:issue]
api-token = <YOUR_API_TOKEN_HERE>
enterprise-url = <YOUR_GITLAB_INSTANCE_URL>
category = issue
raw_index = gitlab_issues_demo_raw
enriched_index = gitlab_issues_demo_enriched
no-archive = true
sleep-for-rate = true

[gitlab:merge]
api-token = <YOUR_API_TOKEN_HERE>
enterprise-url = <YOUR_GITLAB_INSTANCE_URL>
category = merge_request
raw_index = gitlab_merges_demo_raw
enriched_index = gitlab_merges_demo_enriched
no-archive = true
sleep-for-rate = true
```

This blog post, [GrimoireLab to measure organization’s private development,](https://blog.bitergia.com/2020/06/25/grimoirelab-to-measure-organizations-private-development/) would be much helpful. Let me know if it solves your issue. Please send your configurations if you have any more doubts.

> And will the Chinese interface be launched？

Sorry, I didn't get your question.

Also, we would advise your to use the [docker-compose solution](https://github.com/chaoss/grimoirelab/tree/master/docker-compose) as the docker image is quite outdated. Let us know if you need more help.

**Comment by No6-start** [comment]
setup.cfg
`
[es_collection]
url = http://0.0.0.0:9200

[es_enrichment]
url = http://0.0.0.0:9200
autorefresh = true

[general]
debug = True
logs_dir = /app/log
short_name = test
update = False

[panels]
kibiter_url = http://admin:admin@localhost:5601
gitlab-issues = true
gitlab-merges = true

[phases]
collection = True
enrichment = True
identities = True
panels = True

[sortinghat]
affiliate = True
autoprofile = [github, pipermail, git]
database = test
host = 127.0.0.1
matching = [email]
user = root
password = 
sleep_for = 3600
unaffiliated_group = Unknown


[gitlab:issue]
api-token = y75Gg6t_P51zatkP5WM-
enterprise-url = http://192.168.188.74/
category = issue
raw_index = gitlab_issues_demo_raw
enriched_index = gitlab_issues_demo_enriched
no-archive = true
sleep-for-rate = true

[gitlab:merge]
api-token = y75Gg6t_P51zatkP5WM-
enterprise-url = http://192.168.188.74/
category = merge_request
raw_index = gitlab_merges_demo_raw
enriched_index = gitlab_merges_demo_enriched
no-archive = true
sleep-for-rate = true`

**Comment by No6-start** [comment]
projects.json
`{
    "Chaoss": {
	
		"git": [
			"http://192.168.188.74/config/dysub.git"
		]
		
        "gitlab:issue": [
            "http://192.168.188.74/config/dysub",
        ],
        "gitlab:merge": [
            "http://192.168.188.74/config/dysub",
        ]
    }
}
`

**Comment by No6-start** [comment]
These are my configuration files. Please evaluate whether they are correct.

I hope I can use docker image to complete the job. Where should I map the configuration file?

**Comment by vchrombie** [comment]
Hi @No6-start 

Your setup.cfg file looks good except that the `git` configurations are missing. You need to add them too if you want the `git` data source.
```
[git]
raw_index = git_raw
enriched_index = git_enriched
latest-items = true
```
Please note that I haven't added any studies configurations, you can find the complete documentation about it from [chaoss/grimoirelab-sirmordred](https://github.com/chaoss/grimoirelab-sirmordred#contents).

The projects.json file has some formatting errors. You are missing a comma after the `git` data source. The contents of the file should be a valid JSON file, please use any [JSON Formatter & Validator](https://jsonformatter.curiousconcept.com/) if you are doubtful.
```
{
    "Chaoss":{
        "git":[
            "http://192.168.188.74/config/dysub.git"
        ],
        "gitlab:issue":[
            "http://192.168.188.74/config/dysub"
        ],
        "gitlab:merge":[
            "http://192.168.188.74/config/dysub"
        ]
    }
}
```

> I hope I can use docker image to complete the job, where should I map the configuration file, and what is the latest docker image command?

As said already, the docker solution is quite outdated (the image was updated a year back). The [docker-compose solution](https://github.com/chaoss/grimoirelab/tree/master/docker-compose) is also similar to the docker image and it pulls the latest changes too. So, I would suggest you use this. I'm afraid if you will also end up with some errors because of the outdated image like in the issue https://github.com/chaoss/grimoirelab/issues/412.

The documentation in the [grimoirelab/docker-compose](https://github.com/chaoss/grimoirelab/tree/master/docker-compose) is quite clear and updated too. You can find everything over there. Also, this blog post should be helpful, [GrimoireLab Install - CHAOSS](https://chaoss.community/blog-post/2019/10/08/simplifying-how-to-install-grimoirelab/).

The configuration files should be changed in the [grimoirelab/default-grimoirelab-settings](https://github.com/chaoss/grimoirelab/tree/master/default-grimoirelab-settings) if you are using the docker-compose solution. You can find the required files in the [docker](https://github.com/chaoss/grimoirelab/tree/master/docker) folder itself if you still want to stick with docker.

**Comment by No6-start** [comment]
Can I not add the volumes section in docker compose, because we haven't modified it. I think I just need to add it setup.conf and projects.json Just fine


      volumes:
        - ../default-grimoirelab-settings/setup.cfg:/home/bitergia/conf/setup.cfg
        - ../default-grimoirelab-settings/aliases.json:/home/bitergia/conf/aliases.json
        - ../default-grimoirelab-settings/projects.json:/home/bitergia/conf/projects.json
        - ../default-grimoirelab-settings/organizations.json:/home/bitergia/conf/organizations.json
        - ../default-grimoirelab-settings/identities.yml:/home/bitergia/conf/identities.yml
        - /tmp/:/home/bitergia/logs

**Comment by vchrombie** [comment]
Hi @No6-start 

> Can I not add the volumes section in docker compose, because we haven't modified it. I think I just need to add it setup.conf and projects.json Just fine

You don't need to change anything in the [docker-compose.yml](https://github.com/chaoss/grimoirelab/blob/master/docker-compose/docker-compose.yml) file. You just need to update the setup.cfg and projects.json files present in the [default-grimoirelab-settings](https://github.com/chaoss/grimoirelab/tree/master/default-grimoirelab-settings) folder.

**Comment by No6-start** [comment]
I successfully deployed dokcer compose, but it seems that there is something wrong with my configuration, because there is no data after a long time. I don't know where my problem is and how to eliminate it. I didn't find the log  o(╥﹏╥)o

**Comment by vchrombie** [comment]
Hi @No6-start, sorry to hear that.

> I successfully deployed dokcer compose, but it seems that there is something wrong with my configuration, because there is no data after a long time. I don't know where my problem is and how to eliminate it.

Can you please share the configurations so that we can have a closer look?

> I didn't find the log o(╥﹏╥)o

The logs are stored at `/home/bitergia/logs` as defined in the [docker-compose.yml#L56](https://github.com/chaoss/grimoirelab/blob/master/docker-compose/docker-compose.yml#L56) file.

Make sure the docker user has read and write permission on that directory `/tmp/` in this case, but you can create another directory. If you don't know which user is using your docker, create a new directory and assign it `777` and also need to change the docker compose file.

example:
```
$ mkdir ~/mordred-log
$ chmod 777 ~/mordred-log
```
docker-compose.yml
```
    mordred:
      restart: on-failure:5
      image: bitergia/mordred:latest
      volumes:
        ...
        ...
        - /home/NAME/mordred-log:/home/bitergia/logs
      depends_on:
```
After these changes, you need to restart the docker container, [help](https://github.com/chaoss/grimoirelab/tree/master/docker-compose#common-questions).

Ideally, you need not change anything other than the `setup.cfg` and `projects.json` in the [default-grimoirelab-settings](https://github.com/chaoss/grimoirelab/tree/master/default-grimoirelab-settings) folder.

Also it looks like it's a private instance of GitLab so there maybe a case that the token might not be working, so please check if the token is valid.

Hope this helps.
Thanks.

**Comment by No6-start** [comment]
[config.zip](https://github.com/chaoss/grimoirelab/files/6114591/config.zip)


I only modified the setup and project configuration files

**Comment by vchrombie** [comment]
The configurations looks all fine, except for one extra comma in the `git` sources in `projects.json` .

projects.json
```
{
    "grimoirelab": {
      	"meta": {
        	"title": "GrimoireLab"
      	},
      	"git":[
            "http://192.168.188.74/config/dysub.git"
        ],
        "gitlab:issue":[
            "http://192.168.188.74/config/dysub"
        ],
        "gitlab:merge":[
            "http://192.168.188.74/config/dysub"
        ]
    }
 }
```
setup.cfg
```
[general]
short_name = GrimoireLab
update = true
min_update_delay = 60
debug = false
logs_dir = /home/bitergia/logs
aliases_file = /home/bitergia/conf/aliases.json

[projects]
projects_file = /home/bitergia/conf/projects.json

[es_collection]
url = http://elasticsearch:9200

[es_enrichment]
url = http://elasticsearch:9200
autorefresh = true

[sortinghat]
host = mariadb
user = root
password =
database = demo_sh
load_orgs = true
orgs_file = /home/bitergia/conf/organizations.json
autoprofile = [github, pipermail, git]
matching = [email]
sleep_for = 100
unaffiliated_group = Unknown
affiliate = true
strict_mapping = false
reset_on_load = false
identities_file = [/home/bitergia/conf/identities.yml]
identities_format = grimoirelab

[panels]
kibiter_time_from = now-5y
kibiter_default_index = git
kibiter_url = http://kibiter:5601
kibiter_version = 6.1.4-1
gitlab-issues = true
gitlab-merges = true

[phases]
collection = true
identities = true
enrichment = true
panels = true

[git]
raw_index = git_demo_raw
enriched_index = git_demo_enriched
latest-items = true
studies = [enrich_demography:git, enrich_areas_of_code:git, enrich_onion:git]

[gitlab:issue]
api-token = xxxx
enterprise-url = http://192.168.188.74/
raw_index = gitlab_issues_demo_raw
enriched_index = gitlab_issues_demo_enriched
no-archive = true
sleep-for-rate = true

[gitlab:merge]
api-token = xxxx
enterprise-url = http://192.168.188.74/
raw_index = gitlab_merges_demo_raw
enriched_index = gitlab_merges_demo_enriched
no-archive = true
sleep-for-rate = true

## studies based on enriched indexes

[enrich_demography:git]

[enrich_areas_of_code:git]
in_index = git_demo_raw
out_index = git-aoc_demo_enriched

[enrich_onion:git]
in_index = git
out_index = git-onion_demo_enriched
contribs_field = hash
```

Please update the configurations and restart the docker compose to see if it's working. Also please see that it might take 10-15 min time for the data to be collected, enriched and for visualizations to be formed.

If it's still not working, I feel there might be some problem with the token and the gitlab instance. Please make sure you have the correct access rights for the token. We can debug more if you can provide us the logs.

Thanks.

**Comment by No6-start** [comment]
[all.log](https://github.com/chaoss/grimoirelab/files/6139258/all.log)
I checked the log at TMP, but I didn't see any useful error information in the content. I created a new token on the gitlab instance to ensure that there was no problem with the token.

**Comment by No6-start** [comment]
How can I make sure the application is successfully connected to the gitlab instance

**Comment by vchrombie** [comment]
> How can I make sure the application is successfully connected to the gitlab instance

Sorry @No6-start, I didn't understand your query. Can you please explain it more?

**Comment by No6-start** [comment]
There is no content on the grimoirelab page, and no error is found in the log. How can I confirm that I am connected to gitlab

**Comment by No6-start** [comment]
[all.log](https://github.com/chaoss/grimoirelab/files/6139508/all.log)
This is the log in TMP

**Comment by No6-start** [comment]
What should I do？

**Comment by zhquan** [comment]
Hi @No6-start 

You have an error on your `projects.json` file. Check if your file is in valid JSON format. If you are using this https://github.com/chaoss/grimoirelab/issues/413#issuecomment-787714426 the JSON forma is not valid
```
2021-03-15 03:39:50,521 - sirmordred.task_projects - INFO - Reading projects data from  /home/bitergia/conf/projects.json 
2021-03-15 03:39:50,521 - sirmordred.task_manager - ERROR - [Global tasks] Exception in Task Manager Expecting value: line 8 column 9 (char 155)
Traceback (most recent call last):
  File "/usr/local/lib/python3.7/dist-packages/sirmordred/task_manager.py", line 97, in run
    task.execute()
  File "/usr/local/lib/python3.7/dist-packages/sirmordred/task_projects.py", line 133, in execute
    projects = json.load(fprojects)
  File "/usr/lib/python3.7/json/__init__.py", line 296, in load
    parse_constant=parse_constant, object_pairs_hook=object_pairs_hook, **kw)
  File "/usr/lib/python3.7/json/__init__.py", line 348, in loads
    return _default_decoder.decode(s)
  File "/usr/lib/python3.7/json/decoder.py", line 337, in decode
    obj, end = self.raw_decode(s, idx=_w(s, 0).end())
  File "/usr/lib/python3.7/json/decoder.py", line 355, in raw_decode
    raise JSONDecodeError("Expecting value", s, err.value) from None
json.decoder.JSONDecodeError: Expecting value: line 8 column 9 (char 155)
2021-03-15 03:39:50,523 - sirmordred.sirmordred - ERROR - <class 'json.decoder.JSONDecodeError'>
```

I hope it helps you.

Best,
Quan

**Comment by No6-start** [comment]
Thank you. I have modified the contents of the file. At the same time, I see some other errors in the log
projects.json
{
    "grimoirelab": {
      "meta": {
        "title": "GrimoireLab"
      },
      "git":[
            "http://192.168.188.74/config/dysub.git"
        ],
        "gitlab:issue":[
            "http://192.168.188.74/config/dysub"
        ],
        "gitlab:merge":[
            "http://192.168.188.74/config/dysub"
        ]
    }
}

[all (2).log](https://github.com/chaoss/grimoirelab/files/6145540/all.2.log)

**Comment by No6-start** [comment]
I tested my JSON file and found that it was in the correct format, but why did I still report an error

**Comment by zhquan** [comment]
Hi @No6-start 

> 2021-03-16 01:54:02,810 - grimoire_elk.elastic - ERROR - Error creating ES mappings {"error":{"root_cause":[{"type":"cluster_block_exception","reason":"blocked by: [FORBIDDEN/12/index read-only / allow delete (api)];"}],"type":"cluster_block_exception","reason":"blocked by: [FORBIDDEN/12/index read-only / allow delete (api)];"},"status":403}. Mapping:

You don't have enough free space in your machine (elasticsearch change automatically to read only mode when you have less than 5% of free disk space).

When the free space is fixed you have to execute
```
curl -XPUT -H "Content-Type: application/json" http://localhost:9200/_all/_settings -d '{"index.blocks.read_only_allow_delete": null}'
```
Or recreate your elasticsearch container if you are using docker

Best,
Quan

**Comment by No6-start** [comment]
Can I modify the mount path used by es, only 7g space under Var

**Comment by No6-start** [comment]
[all.log](https://github.com/chaoss/grimoirelab/files/6154335/all.log)
I modified the default storage path of docker to ensure enough space. I recreated the log file and found some other problems. I need help to have a look.

**Comment by zhquan** [comment]
Hi @No6-start 

> 21-03-17 07:00:29,337 - grimoire_elk.elk - ERROR - Error feeding raw from git (http://192.168.188.74/config/dysub.git): git command - Cloning into bare repository '/home/bitergia/.perceval/repositories/http://192.168.188.74/config/dysub.git-git'...
fatal: could not read Username for 'http://192.168.188.74': No such device or address
Traceback (most recent call last):
  File "/usr/local/lib/python3.7/dist-packages/grimoire_elk/elk.py", line 162, in feed_backend
    ocean_backend.feed(**params)
  File "/usr/local/lib/python3.7/dist-packages/grimoire_elk/raw/elastic.py", line 228, in feed
    self.feed_items(items)
  File "/usr/local/lib/python3.7/dist-packages/grimoire_elk/raw/elastic.py", line 244, in feed_items
    for item in items:
  File "/usr/local/lib/python3.7/dist-packages/perceval/backend.py", line 226, in fetch
    for item in self.fetch_items(category, **kwargs):
  File "/usr/local/lib/python3.7/dist-packages/perceval/backends/core/git.py", line 161, in fetch_items
    latest_items, no_update)
  File "/usr/local/lib/python3.7/dist-packages/perceval/backends/core/git.py", line 274, in __fetch_from_repo
    repo = self.__create_git_repository()
  File "/usr/local/lib/python3.7/dist-packages/perceval/backends/core/git.py", line 325, in __create_git_repository
    repo = GitRepository.clone(self.uri, self.gitpath)
  File "/usr/local/lib/python3.7/dist-packages/perceval/backends/core/git.py", line 833, in clone
    cls._exec(cmd, env=env)
  File "/usr/local/lib/python3.7/dist-packages/perceval/backends/core/git.py", line 1338, in _exec
    raise RepositoryError(cause=cause)
perceval.errors.RepositoryError: git command - Cloning into bare repository '/home/bitergia/.perceval/repositories/http://192.168.188.74/config/dysub.git-git'...
fatal: could not read Username for 'http://192.168.188.74': No such device or address

Make sure the URL of the repo is correct and you have permission to clone it
Try to do this:
1. enter to the mordred container `docker exec -it MORDRED_CONTAINER_NAME /bin/bash`
2. execute perceval `perceval git http://192.168.188.74/config/dysub.git`
3. try to clone `git clone http://192.168.188.74/config/dysub.git`

Best,
Quan

**Comment by No6-start** [comment]
Mordred_ CONTAINER_ NAME is grimoirelab/hatstall:lates？

**Comment by No6-start** [comment]
I'm in grimoirelab/ hatstall:latest You can clone the warehouse normally, but you need to enter the account and password
if Mordred_ CONTAINER_ NAME is grimoirelab/hatstall:lates

**Comment by zhquan** [comment]
@No6-start 
> grimoirelab/hatstall:lates

That is `hatstall` container https://github.com/chaoss/grimoirelab/blob/master/docker-compose/docker-compose.yml#L12

This is the `mordred` container https://github.com/chaoss/grimoirelab/blob/master/docker-compose/docker-compose.yml#L47

> I'm in grimoirelab/ hatstall:latest You can clone the warehouse normally, but you need to enter the account and password

Then you have to put your credentials/token on the projecs.json. You have to be able to clone it directly.
-  `http://USER:PASS@192.168.188.74/config/dysub.git`
-  `http://USER:TOKEN@192.168.188.74/config/dysub.git` (you only need read permission)

> I didn't find the Perceval command

Only the `mordred` container has `perceval`

**Comment by No6-start** [comment]
Thank you. The data has appeared on the page, but it seems different from my total warehouse data. I have 90 merges in total, but the statistics on the page are 3 merges

I want to know when I use http://USER:PASS@192.168.188.74/config/dysub.git
Does the app count all branches?

**Comment by No6-start** [comment]
I have multiple warehouses and branches to check. How should I configure them? In the following example, I run it with errors

{
    "dysub": {
      "meta": {
        "title": "midl-dysub"
      },
      "git":[
            "http://y75Gg6t_P51zatkP5WM-:y75Gg6t_P51zatkP5WM-@git.1an.com/midl/dysub.git"
        ],
        "gitlab:issue":[
            "http://y75Gg6t_P51zatkP5WM-:y75Gg6t_P51zatkP5WM-@git.1an.com/midl/dysub"
        ],
        "gitlab:merge":[
            "http://y75Gg6t_P51zatkP5WM-:y75Gg6t_P51zatkP5WM-@git.1an.com/midl/dysub"
        ]
    },
    "nclm": {
      "meta": {
        "title": "core-nclm"
      },
      "git":[
            "http://y75Gg6t_P51zatkP5WM-:y75Gg6t_P51zatkP5WM-@git.1an.com/core/nclm.git"
        ],
        "gitlab:issue":[
            "http://y75Gg6t_P51zatkP5WM-:y75Gg6t_P51zatkP5WM-@git.1an.com/core/nclm"
        ],
        "gitlab:merge":[
            "http://y75Gg6t_P51zatkP5WM-:y75Gg6t_P51zatkP5WM-@git.1an.com/core/nclm"
        ]
    }
}

**Comment by zhquan** [comment]
Hi @No6-start 

You only need to add the credentials in `git`. For `gitlab:issue` and `gitlab:merge` you have to add your `TOKEN` on the `setup.cfg`.
```
[gitlab:issue]
api-token = TOKEN

[gitlab:merge]
api-token = TOKEN
```

Remove your credentials for `issues` and `merges` on the `projects.json` and remove your indexes if they are wrong and Mordred will create them from scratch. If you are using https://github.com/chaoss/grimoirelab/issues/413#issuecomment-787714069 you have to remove 
- giltab_issues_demo_raw
- gitlab_issues_demo_enriched
- gitlab_merges_demo_raw
- gitlab_merges_demo_enriched

On kibana go to `dev tools`
```
DELETE <index>
```

Best,
Quan

**Comment by No6-start** [comment]
I found that the index data didn't match the actual data,
At the same time, I have multiple warehouses and branches to check. How should I configure them? In the following example, I made an error

**Comment by zhquan** [comment]
@No6-start - git fetch commit from all branches so you don't have to do anything. But if you want to know which commits belong to which branch, you have to activate the study `enrich_git_branches` see https://github.com/chaoss/grimoirelab-sirmordred#git-

By default, this study will execute once per week [7, 14, 21, 28] https://github.com/chaoss/grimoirelab-elk/blob/master/grimoire_elk/enriched/git.py#L850 so if you want to execute it today add this
```
[enrich_git_branches:git]
run_month_days = [19]
```

Note that if `latest-items` is activated, only new commits will be fetched even if you remove the index. If you want to fetch them from scratch you must deactivate it.

```
[git]
latest-items = false
```

**Comment by No6-start** [comment]
I have multiple projects. How to write JSON？

**Comment by vchrombie** [comment]
> I have multiple projects. How to write JSON？

I understand you are asking how to configure multiple projects. You can add them in the projects.json file.
```
{
    "grimoirelab": {
      	"git":[
            "http://192.168.188.74/config/one.git",
            "http://192.168.188.74/config/two.git",
            "http://192.168.188.74/config/three.git"
        ]
    }
 }
```
You can read more about it from the [documentation](https://github.com/chaoss/grimoirelab-sirmordred#projectsjson-).

**Comment by No6-start** [comment]
Can I use this style because I want to manage multiple groups easily
{
    "grimoirelab": {
      	"git":[
            "http://192.168.188.74/config/one.git",
        ]
      	"git":[
            "http://192.168.188.74/config/two.git",
        ]
    }
    "grimoirelab-tow": {
      	"git":[
            "http://192.168.188.74/config/three.git",
        ]
    }
 }

**Comment by No6-start** [comment]
Is there only one JSON style?

**Comment by vchrombie** [comment]
> Is there only one JSON style?

The JSON should be of the same format mentioned in the [documentation](https://github.com/chaoss/grimoirelab-sirmordred#projectsjson-).
You can [JSON Formatter & Validator](jsonformatter.curiousconcept.com/) for avoiding any typos or formatting mistakes.

**Comment by No6-start** [comment]
Can I use this style?
{
“ grimoirelab”：{
“ git”：[
“ http://192.168.188.74/config/one.git”,
“ http://192.168.188.74/config/two.git”,
]
}
“ grimoirelab-tow”：{
“ git”：[
“ http://192.168.188.74/config/three.git”,
]
}
}

**Comment by No6-start** [comment]
I've tested some content, and now there are a few questions.
1. I turned on enrich_ git_ Branches, but you don't see the contents of the branches
2. According to project.json The contents in the file do not show merge and issue, but only commit.

**Comment by vchrombie** [comment]
Hi @No6-start 

The projects.json file looks fine to me. If the data is still not visible, there must be a problem with the setup.cfg file. It would be great if you can share the sanitized version of it.
```
{
    "dysub":{
        "meta":{
            "title":"midl-dysub"
        },
        "git":[
            "http://phU-DaG9zSKGTvsz5BH4:phU-DaG9zSKGTvsz5BH4@git.1an.com/midl/dysub.git",
            "http://phU-DaG9zSKGTvsz5BH4:phU-DaG9zSKGTvsz5BH4@git.1an.com/core/nclm.git"
        ],
        "gitlab:issue":[
            "http://git.1an.com/midl/dysub",
            "http://git.1an.com/core/nclm"
        ],
        "gitlab:merge":[
            "http://git.1an.com/midl/dysub",
            "http://git.1an.com/core/nclm"
        ]
    }
}
```

>  3.setup.cfg Gitlab in_ issues_ demo_ raw、gitlab_ issues_ demo_ enriched、gitlab_ merges_ demo_ raw、gitlab_ merges_ demo_ Does enriched use the default value? I don't know what these values mean

These are the indexes where the elasticsearch documents are stored. Earlier you had the wrong configuration, so please delete them so that Mordred will start from scratch. Please check this https://github.com/chaoss/grimoirelab/issues/413#issuecomment-802661116.

**Comment by No6-start** [comment]
The attachment is mine setup.cfg , I deleted in_ issues_  demo_  raw、gitlab_  issues_  demo_  enriched、gitlab_  merges_  demo_  raw、gitlab_  merges_  Demo, but the page displays an error:
Couldn't find any Elasticsearch data
You'll need to index some data into Elasticsearch before you can create an index pattern. Learn how or get started with some sample data sets.


[setup.txt](https://github.com/chaoss/grimoirelab/files/6210493/setup.txt)

**Comment by vchrombie** [comment]
Hi @No6-start, I see that some of the configurations are missing, including the gitlab index names, from the setup.cfg file. 

I added back them here, please check it and let us know.
```
[general]
short_name = GrimoireLab
update = true
min_update_delay = 60
debug = false
logs_dir = /home/bitergia/logs
aliases_file = /home/bitergia/conf/aliases.json

[projects]
projects_file = /home/bitergia/conf/projects.json

[es_collection]
url = http://elasticsearch:9200

[es_enrichment]
url = http://elasticsearch:9200
autorefresh = true

[sortinghat]
host = mariadb
user = root
password =
database = demo_sh
load_orgs = true
orgs_file = /home/bitergia/conf/organizations.json
autoprofile = [github, pipermail, git]
matching = [email]
sleep_for = 100
unaffiliated_group = Unknown
affiliate = true
strict_mapping = false
reset_on_load = false
identities_file = [/home/bitergia/conf/identities.yml]
identities_format = grimoirelab

[panels]
kibiter_time_from = now-5y
kibiter_default_index = git
kibiter_url = http://kibiter:5601
kibiter_version = 6.1.4-1
gitlab-issues = true
gitlab-merges = true

[phases]
collection = true
identities = true
enrichment = true
panels = true

[enrich_git_branches:git]
run_month_days = [25]

[git]
raw_index = git_demo_raw
enriched_index = git_demo_enriched
latest-items = false
studies = [enrich_demography:git, enrich_git_branches:git, enrich_areas_of_code:git, enrich_onion:git]

[gitlab:issue]
api-token = phU-DaG9zSKGTvsz5BH4
enterprise-url = http://git.1an.com/
raw_index = gitlab_issues_demo_raw
enriched_index = gitlab_issues_demo_enriched
no-archive = true
sleep-for-rate = true

[gitlab:merge]
api-token = phU-DaG9zSKGTvsz5BH4
enterprise-url = http://git.1an.com/
raw_index = gitlab_merges_demo_raw
enriched_index = gitlab_merges_demo_enriched
no-archive = true
sleep-for-rate = true

## studies based on enriched indexes

[enrich_demography:git]

[enrich_areas_of_code:git]
in_index = git_demo_raw
out_index = git-aoc_demo_enriched

[enrich_onion:git]
in_index = git
out_index = git-onion_demo_enriched
contribs_field = hash
```

**Comment by No6-start** [comment]
[gir.zip](https://github.com/chaoss/grimoirelab/files/6218904/gir.zip)
1. The attachment is my configuration and log. I found that only git exists on the page, without merge and issue.
2. And I found that the latest submission record was on March 4, and I submitted it every day in March.

**Comment by vchrombie** [comment]
Hi @No6-start, I'm sorry for the inconvenience caused.

I missed `category` configuration for the `gitlab:issue` and `gitlab:merge` in the https://github.com/chaoss/grimoirelab/issues/413#issuecomment-808138649. I think this is the reason why the data is not being fetched. Sorry, my bad.

setup.cfg
```
[gitlab:issue]
category = issue
api-token = xxxx
enterprise-url = http://git.1an.com/
raw_index = gitlab_issues_demo_raw
enriched_index = gitlab_issues_demo_enriched
no-archive = true
sleep-for-rate = true

[gitlab:merge]
category = merge_request
api-token = xxxx
enterprise-url = http://git.1an.com/
raw_index = gitlab_merges_demo_raw
enriched_index = gitlab_merges_demo_enriched
no-archive = true
sleep-for-rate = true
```

Can you update the setup.cfg file and restart the docker-compose? Make sure you run the `sysctl -w vm.max_map_count=262144` command to provide sufficient virtual memory to the es.

**Comment by No6-start** [comment]
Sorry, I rebuilt docker compose, but I still don't have merge and issue content
![20210329112745](https://user-images.githubusercontent.com/54304180/112783392-ced24600-9081-11eb-9b98-e1e320b01582.png)

**Comment by vchrombie** [comment]
Hi @No6-start, can you connect at `#grimoirelab` at [Freenode](http://freenode.net/) for a quick chat?

**Comment by No6-start** [comment]
> 嗨，@ No6-start，您可以`#grimoirelab`在[Freenode](http://freenode.net/)进行连接以进行快速聊天吗？

yes

**Comment by No6-start** [comment]
> 嗨，@ No6-start，您可以`#grimoirelab`在[Freenode](http://freenode.net/)进行连接以进行快速聊天吗？

I can try

**Comment by vchrombie** [comment]
> yes

I will be available for the next 4 hours.
Thanks.

**Comment by No6-start** [comment]
> > 是的
> 
> 接下来的4个小时我会上班。
> 谢谢。

How can I join your channel in freenode

**Comment by No6-start** [comment]
Two hours later, I saw the merge data. I think the configuration is correct. Thank you

**Comment by No6-start** [comment]
Two hours later, I saw the merge data. I think the configuration is correct. Thank you. Can merge be divided into multiple branches

**Comment by No6-start** [comment]
Have you considered adding another language version?

**Comment by vchrombie** [comment]
Hi @No6-start 

> Two hours later, I saw the merge data, and I think the configuration is correct, thank you. 

Oh, that's great news! 
It would be great to close this issue if your query is resolved. You can open another one later if needed.

> When I need to view other branches, do I need to enter the content in the search?

Right now, this information is available only in the raw index (`source_branch` and `target_branch`), but not propagated to the enriched index.

See below a Perceval document generated by the gitlab backend (category: merge_request). 
```
$ perceval gitlab gitlab-org gitlab-git-test --category merge_request -t xxxx
{
    "backend_name":"GitLab",
    "backend_version":"0.12.0",
    "category":"merge_request",
    "classified_fields_filtered":null,
    "data":{
        "approvals_before_merge":null,
        "assignee":{
            "avatar_url":"https://assets.gitlab-static.net/uploads/-/system/user/avatar/280339/avatar.png",
            "id":280339,
            "name":"Rub\u00e9n D\u00e1vila",
            "state":"active",
            "username":"rdavila",
            "web_url":"https://gitlab.com/rdavila"
        },
        ...
        ...
        ...
        "pipeline":null,
        "project_id":83282,
        "reference":"!10",
        "references":{
            "full":"gitlab-org/gitlab-git-test!10",
            "relative":"!10",
            "short":"!10"
        },
        "reviewers":[
            
        ],
        "sha":"c54ad072fabee9f7bf9b2c6c67089db97ebfbecd",
        "should_remove_source_branch":null,
        "source_branch":"rd-add-file-larger-than-1-mb", <--------
        "source_project_id":83282,
        "squash":false,
        "squash_commit_sha":null,
        "state":"closed",
        "subscribed":false,
        "target_branch":"master", <--------
        "target_project_id":83282,
        "task_completion_status":{
            "completed_count":0,
            "count":0
        },
        "time_stats":{
            "human_time_estimate":null,
            "human_total_time_spent":null,
            "time_estimate":0,
            "total_time_spent":0
        },
        "title":"Add file larger than 1 mb",
        ...
        ...
        "web_url":"https://gitlab.com/gitlab-org/gitlab-git-test/-/merge_requests/10",
        "work_in_progress":false
    },
    "origin":"https://gitlab.com/gitlab-org/gitlab-git-test",
    "perceval_version":"0.17.3",
    "search_fields":{
        "groups":null,
        "iid":10,
        "item_id":"8960429",
        "owner":"gitlab-org",
        "project":"gitlab-git-test"
    },
    "tag":"https://gitlab.com/gitlab-org/gitlab-git-test",
    "timestamp":1617017616.074928,
    "updated_on":1523247338.428,
    "uuid":"feadeb5b198cf62ead5bc1b222df92f17021582f"
}
```

The gitlab enriched handles this data, https://github.com/chaoss/grimoirelab-elk/blob/master/grimoire_elk/enriched/gitlab.py.

> Have you considered adding another language version?

I don't think know if the community has planned about this and how far is this feasible. In any case, please feel free to open another issue to discuss this.

Thanks.

**Comment by No6-start** [comment]
Sorry, I didn't understand the example of Perceval document. I didn't find the relevant JSON rules on the page

**Comment by vchrombie** [comment]
Sorry for the late reply, @No6-start 

> Sorry, I didn't understand the example of Perceval document. I didn't find the relevant JSON rules on the page

Perceval is the grimoirelab component that retrieves data from the data sources like GitLab. You can learn more about it from the [chaoss/grimoirelab-perceval](https://github.com/chaoss/grimoirelab-perceval) repository.

You can find the example for running the [gitlab perceval backend](https://github.com/chaoss/grimoirelab-perceval#gitlab) to fetch `issue` and `merge_request` data. You can specify either `issue` or `merge_request` using the `--category` flag.

example for fetching `merge_request` data from the https://gitlab.com/gitlab-org/gitlab-git-test/ project using perceval
```
$ perceval gitlab gitlab-org gitlab-git-test --category merge_request -t xxxx
```

These Perceval documents are later stored as raw and enriched indexes by [chaoss/grimoirelab-elk](https://github.com/chaoss/grimoirelab-elk/) in ElasticSearch and Kibiter shows the visualizations.
https://jsmanrique.github.io/grimoirelab-user-guide/docs/components/

It would be better if we close this issue since it is unrelated to the conversation and the issue has been resolved. 
Please feel free to open a new issue if you have any doubts. 

Thanks.

**Comment by vchrombie** [comment]
@No6-start
Since the issue is resolved, it would be great if you can close this issue.

**Comment by zhquan** [comment]
I close the issue
